Sources/AsyncHTTPClient/ConnectionPool/HTTPRequestStateMachine.swift

```diff
@@ -0,0 +1,625 @@
+            )
+
+            self.state = .running(requestState, responseState)
+
+            return .sendBodyPart(part)
+
+        case .failed:
+            return .wait
+
+        case .finished:
+            // A request may be finished, before we have send all parts. This might be the case if
+            // the server responded with an HTTP status code that is equal or larger to 300
+            // (Redirection, Client Error or Server Error). In those cases we pause the request body
+            // stream as soon as we have received the response head and we succeed the request as
+            // when response end is received. This may mean, that we succeed a request, even though
+            // we have not sent all it's body parts.
+
+            // We may still receive something, here because of potential race conditions with the
+            // producing thread.
+            return .wait
+        }
+    }
+
+    mutating func requestStreamFinished() -> Action {
+        switch self.state {
+        case .initialized,
+             .verifyingRequest,
+             .waitForChannelToBecomeWritable,
+             .running(.endSent, _),
+             .finished:
+            preconditionFailure("A request body stream end is only expected if we are in state request streaming. Invalid state: \(self.state)")
+
+        case .running(.streaming(let expectedBodyLength, let sentBodyBytes, _), .waitingForHead):
+            if let expected = expectedBodyLength, expected != sentBodyBytes {
+                let error = HTTPClientError.bodyLengthMismatch
+                self.state = .failed(error)
+                return .failRequest(error, .close, clearReadTimeoutTimer: false)
+            }
+
+            self.state = .running(.endSent, .waitingForHead)
+            return .sendRequestEnd(startReadTimeoutTimer: self.idleReadTimeout)
+
+        case .running(.streaming(let expectedBodyLength, let sentBodyBytes, _), .receivingBody(let head, let streamState)):
+            assert(head.status.code < 300)
+
+            if let expected = expectedBodyLength, expected != sentBodyBytes {
+                let error = HTTPClientError.bodyLengthMismatch
+                self.state = .failed(error)
+                return .failRequest(error, .close, clearReadTimeoutTimer: self.idleReadTimeout != nil)
+            }
+
+            self.state = .running(.endSent, .receivingBody(head, streamState))
+            return .sendRequestEnd(startReadTimeoutTimer: self.idleReadTimeout)
+
+        case .running(.streaming(let expectedBodyLength, let sentBodyBytes, _), .endReceived):
+            if let expected = expectedBodyLength, expected != sentBodyBytes {
+                let error = HTTPClientError.bodyLengthMismatch
+                self.state = .failed(error)
+                return .failRequest(error, .close, clearReadTimeoutTimer: false)
+            }
+
+            self.state = .finished
+            return .succeedRequest(.none, clearReadTimeoutTimer: false)
```

Comment:
Yeah, I don't know why the read timeout isn't getting cleared here. I think whether the read timeout is cleared or reset should not be the responsibility of this state machine: the read timeout should get set when either we've sent our `.end` or received our first `.head`, and from that point on it shouldn't be our problem. It adds too many responsibilities to this object.

Reply:
Alternatively, if you _really_ want it to be composed into this state machine, factor it out into a little state machine of its own and give it an interface that you can call from here. It'll make it much easier to validate its correctness. But honestly I think it's overcomplicating this object.